src/manifold/executor.clj

```diff
@@ -32,25 +32,42 @@
        (finally
          (.set executor-thread-local executor#)))))
 
+(defn- ^Thread new-thread
+  ([group target name]
+   (Thread. group target name))
+  ([group target name stack-size]
+   (Thread. group target name stack-size)))
+
 (defn ^ThreadFactory thread-factory
+  "Returns a `java.util.concurrent.ThreadFactory`.
+
+   |:---|:----
+   | `name-generator` | a zero-argument function, which, when invoked returns the name of the `java.lang.Thread` that will be created. |
+   | `executor-promise` | a promise eventually containing a `java.util.concurrent.Executor` that will be stored on `manifold.executor/executor-thread-local`. |
+   | `stack-size` | the desired stack size for the new thread, or nil/zero to indicate that this parameter is to be ignored. |
+   | `daemon?` | marks the created threads as either daemon or user threads. The Java Virtual Machine exits when the only threads running are all daemon threads. |
+   | `new-thread-fn` | a three/four arguments function which returns an implementation of `java.lang.Thread` when called. |"
   ([name-generator executor-promise]
-   (thread-factory name-generator executor-promise nil true))
+   (thread-factory name-generator executor-promise nil true nil))
   ([name-generator executor-promise stack-size]
-   (thread-factory name-generator executor-promise stack-size true))
+   (thread-factory name-generator executor-promise stack-size true nil))
   ([name-generator executor-promise stack-size daemon?]
-   (reify ThreadFactory
-     (newThread [_ runnable]
-       (let [name        (name-generator)
-             curr-loader (.getClassLoader (class thread-factory))
-             f           #(do
-                            (.set executor-thread-local @executor-promise)
-                            (.run ^Runnable runnable))]
-         (doto
-           (if stack-size
-             (Thread. nil f name stack-size)
-             (Thread. nil f name))
-           (.setDaemon daemon?)
-           (.setContextClassLoader curr-loader)))))))
+   (thread-factory name-generator executor-promise stack-size daemon? nil))
+  ([name-generator executor-promise stack-size daemon? new-thread-fn]
+   (let [new-thread (or new-thread-fn new-thread)]
+     (reify ThreadFactory
+       (newThread [_ runnable]
+         (let [name        (name-generator)
+               curr-loader (.getClassLoader (class thread-factory))
+               f           #(do
+                              (.set executor-thread-local @executor-promise)
+                              (.run ^Runnable runnable))
+               thread      (if stack-size
+                             ^Thread (new-thread nil f name stack-size)
```

Comment:
Maybe we could enforce the usage of a four arguments function here by coercing the `stack-size` to 0. [1]
What do you think?

[1] : https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/Thread.java#L624

Reply:
That's fine, as long we also document that 0 means "ignore this setting"

Reply:
#cc415a8